### ISSUE TITLE
Rendering performance issue in BitDropdown with search input and large number of BitDropdownItems

### ISSUE THREAD

**Issue by ShahryarSaljoughi** [bitfoundation/bitplatform#9021]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Typing into the search input of `BitDropdown` is slow.
The issue is similar to #7682 .
I believe this is due to the number of items in the `Items` parameter passed to the `BitDropdown` component. 

I've compared the dropdown in different tenants of my multi tenant SaaS. The tenant with only 3 items in the dropdown did not experience the issue, input is laggy if number of dropdown items is 500.
Each key press in search input results in **285ms** of _scripting_ (the term the browser uses) time.  

Currently, I have no idea how to tackle such issues, but I'm glad to be assigned this task.

### Expected Behavior

I expect to have a smooth typing in the search input, meaning that the text is typed as I press keys on keyboard without any lag. I expect the size of Items not to affect the typing experience.

### .NET Version

8.0.300

**Comment by msynk** [comment]
Thanks for contacting us. We're investigating this issue.
We'll let you know if it's possible to work on this issue.

**Comment by msynk** [comment]
@ShahryarSaljoughi isn't this a duplicate of #8579?
if yes I'm going to close this.

**Comment by msynk** [comment]
btw, did you test the Virtualize parameter for large lists?

**Comment by ShahryarSaljoughi** [comment]
Yes I did, didn't help.

**Comment by ShahryarSaljoughi** [comment]
Here are what I found:
- The component is **not** laggy, event with large list of `Items`, **if I omit the `OnSearch` parameter**. I use that parameter to fetch a new set of values and pass them to `Items` parameter.
- The component is **not** laggy, If the action passed to `OnSearch` only does simply `await Task.Delay(4000)`.
- The component is laggy, even if handler of `OnSearch` does not modify the list of items passed to `Items` or assign a new list to it, meaning, If it simply calls the API, still component is laggy.


BTW, I saw how `Immediate` makes data binding occur less. I don't think that would help this issue. I have a reason for that: I tested with only one `keypress` and that took a long time too. 

## Trying Thread Pool
I tried to use a thread from thread pool for calling web API:
```
async Task HandleOnSearchAsync(string? term)
{
    await Task.Run(async () =>
    {
        var pagedResult = await ProductDataProvider.GetProductsAsync(new GetProductsRequest { Pagination = GetProductsRequest.Default.Pagination, SearchTerm = term });
        //Products = pagedResult.Items.ToList();
    });
}
```
Interesting thing happened:  
- On first key press, it is so fast: instantly I see the character typed into the input
- If I wait 2 seconds before typing again, the same fast typing occurs
- If I write two chars, the first one is rendered but it takes up to two seconds for the second character to appear on screen.

## My Conclusion is that
The time consumed for fetching more data from server, causes the lag.

**Comment by msynk** [comment]
thanks for your thorough investigations. it clears a lot of things.
One important aspect of the problem is that this issue only happens when using the ItemProvider feature. is this correct?

**Comment by ShahryarSaljoughi** [comment]
No, I've not used the ItemsProvider feature. To be honest I had not noticed ItemsProvider feature exists.  

Here is a summary of my setup explained in two steps:
**1.** `Dropdown` is a wrapper around `BitDropdown` component.
`Dropdown.razor`:
```
<BitDropdown TItem="BitDropdownItem<TItem>" TValue="TItem"
             Items="@Items.Select(b => new BitDropdownItem<TItem>
                    {
                        Value = b,
                        Text = TitleSelector(b),
                        Class = "item-cls"
                        }).ToList()"
             Virtualize
             ShowSearchBox="@ShowSearchBox"
             SearchBoxPlaceholder="جست و جو کنید"
             OnSearch="HandleSearchTermChanged"
             @bind-Value:get="Value"
             @bind-Value:set="ValueChanged">

</BitDropdown>
```
`Dropdown.razor.cs`:
```
    public async Task HandleSearchTermChanged(string? term)
    {
        if (OnSearch.HasDelegate)
        {
            await OnSearch.InvokeAsync(term);
        }
    }
```
**2.** `ProductSelector` is a component using the `Dropdown` Component
- A list of objects of type `ProductDto` are passed to `Items` parameter.
- `OnSearch` is passed a delegate which fetches a new list objects and passes the new list to `Items` parameter. 
`ProductSelector.razor`:
```
    <Dropdown TItem="ProductDto"
              Items="@(Products)"
              TitleSelector="p => p.Name"
              Placeholder="انتخاب کنید" 
              Label="نام محصول"
              ShowSearchBox
              OnSearch="HandleOnSearchAsync"       
              @bind-Value:get="Product"
              @bind-Value:set="ProductChanged"/>
```
`ProductSelector.razor.cs`:
```
    async Task HandleOnSearchAsync(string? term)
    {
var pagedResult = await ProductDataProvider.GetProductsAsync(new GetProductsRequest { Pagination = GetProductsRequest.Default.Pagination, SearchTerm = term });
Products = pagedResult.Items.ToList();
    }
```

**Comment by ShahryarSaljoughi** [comment]
@msynk Dear Saleh, I found the root cause. The root cause is not related to the Dropdown component.  
I'm really grateful to your efforts and time you dedicated to this issue. I'm sorry about your time and effort on this issue.

For documentation purposes, and helping others facing the same issue, I'll explain the root cause in short:
I have a naive caching mechanism, in order to reduce networking time. However as elaborated [here:](https://stackoverflow.com/questions/61864285/blazor-ui-freeze-even-when-task-is-used#answer-61865580) 

> Any 'heavy operation' will block the UI, that is the expected behaviour.

I solved the problem by eliminating the heavy operations which were eating the UI thread's time.
